nx/source/services/ncm.c

```diff
@@ -1,1029 +1,484 @@
-
-    struct {
-        u64 magic;
-        u64 cmd_id;
-    } *raw;
-    
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-    
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 15;
-    
-    Result rc = serviceIpcDispatch(&db->s);
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-        } *resp = r.Raw;
-
-        rc = resp->result;
-    }
-    
-    return rc;
+    return _ncmCmdNoIo(&db->s, 15);
 }
 
-Result ncmContentMetaDatabaseGetAttributes(NcmContentMetaDatabase* db, const NcmMetaRecord* record, u8* out) {
-    IpcCommand c;
-    ipcInitialize(&c);
+Result ncmContentMetaDatabaseHasContent(NcmContentMetaDatabase* db, bool* out, const NcmContentMetaKey* key, const NcmNcaId* content_id) {
+    const struct {
+        NcmNcaId content_id;
+        NcmContentMetaKey key;
+    } in = { *content_id, *key };
+    return serviceDispatchInOut(&db->s, 16, in, *out);
+}
 
-    struct {
-        u64 magic;
-        u64 cmd_id;
-        NcmMetaRecord meta_record;
-    } *raw;
-    
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-    
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 18;
-    memcpy(&raw->meta_record, record, sizeof(NcmMetaRecord));
-    
-    Result rc = serviceIpcDispatch(&db->s);
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
+Result ncmContentMetaDatabaseListContentMetaInfo(NcmContentMetaDatabase* db, u32* out_entries_written, void* out_meta_info, size_t out_meta_info_size, const NcmContentMetaKey* key, u32 start_index) {
+    const struct {
+        u32 start_index;
+        u32 padding;
+        NcmContentMetaKey key;
+    } in = { start_index, 0, *key };
+    return serviceDispatchInOut(&db->s, 17, in, *out_entries_written,
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out },
+        .buffers = { { out_meta_info, out_meta_info_size } },
+    );
+}
 
-        struct {
-            u64 magic;
-            u64 result;
-            u8 out;
-        } *resp = r.Raw;
+Result ncmContentMetaDatabaseGetAttributes(NcmContentMetaDatabase* db, const NcmContentMetaKey* key, u8* out) {
+    return serviceDispatchInOut(&db->s, 18, *key, *out);
+}
 
-        rc = resp->result;
+Result ncmContentMetaDatabaseGetRequiredApplicationVersion(NcmContentMetaDatabase* db, u64* out_version, const NcmContentMetaKey* key) {
+    if (hosversionBefore(2,0,0)) return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
+    return serviceDispatchInOut(&db->s, 19, *key, *out_version);
+}
 
-        if (R_SUCCEEDED(rc)) {
-            if (out) *out = resp->out;
-        }
-    }
-    
-    return rc;
+Result ncmContentMetaDatabaseGetContentIdByTypeAndIdOffset(NcmContentMetaDatabase* db, NcmNcaId* out_content_id, const NcmContentMetaKey* key, NcmContentType type, u8 id_offset) {
+    if (hosversionBefore(5,0,0)) return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
+    const struct {
+        u32 type;
```

Comment:
Checked sysmodule code, this doesn't seem right?(2 u8s then the struct)

Reply:
Ah yes, it looks like this issue is inherited from ncmContentMetaDatabaseGetContentId having the wrong type (which I fixed), and then based this off of